### PR TITLE
Add long path support for Windows

### PR DESCRIPTION
**Pull request by majaeger**
Currently, ninja only works for paths less than MAX_PATH (255) on Windows. This change makes ninja long-path aware by manifesting the application as "longPathAware", and using the unicode version of windows file APIs

See the below docs for more info:
https://docs.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation
https://docs.microsoft.com/en-us/dotnet/standard/io/file-path-formats#skip-normalization
https://docs.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation#enable-long-paths-in-windows-10-version-1607-and-later

### CODE REVIEW

**Comment by majaeger** [comment]
Appears there are still some issues with chdir

**Comment by majaeger** [comment]
As far as performance, there seems to be negligible difference, though actually the UNICODE is consistently slightly faster. My guess is this is because of the reduced file path verification, where it lets the file driver do more of the work.

On my machine Win10 20H2, for a small/medium sized project, I see:

ANSI: ~51s
UNICODE: ~49s

**Comment by aronyu79** [comment]
Any update on this? Looking forward to it being merged!

**Comment by parbo** [comment]
I would also love for this to be merged. Projects at my workplace are affected by this problem.

**Comment by jonesmz** [comment]
The Linux/test github action will need to succeed before it this PR is merged, most likely.

Also, would the use of C++17's std::filesystem also address this issue? See https://github.com/splinter-build/splinter/tree/16-std-filesystem

File: CMakeLists.txt [hunk 1]
```diff
@@ -205,6 +205,9 @@ if(BUILD_TESTING)
     target_sources(ninja_test PRIVATE src/includes_normalize_test.cc src/msvc_helper_test.cc)
   endif()
   target_link_libraries(ninja_test PRIVATE libninja libninja-re2c)
+  if (MSVC AND "$ENV{VSCMD_ARG_TGT_ARCH}" STREQUAL "x86")
+    target_compile_options(ninja_test PRIVATE "/fsanitize=address")
```
Comment (jonesmz) [inline CMakeLists.txt:209]: Adding /fsanitize=address is not related to adding long path awareness to Ninja.

Don't get me wrong, it should be added to the build, but it should be done as a completely separate PR.

File: debug.log [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[0324/205427.135:ERROR:registration_protocol_win.cc(102)] CreateFile: The system cannot find the file specified. (0x2)
```
Comment (jonesmz) [inline debug.log:1]: There is no reason to add this file to the repository.

File: src/disk_interface.cc [hunk 3]
```diff
@@ -73,13 +90,13 @@ TimeStamp TimeStampFromFileTime(const FILETIME& filetime) {
   return (TimeStamp)mtime - 12622770400LL * (1000000000LL / 100);
 }
 
-TimeStamp StatSingleFile(const string& path, string* err) {
+TimeStamp StatSingleFile(const wstring& path, string* err) {
   WIN32_FILE_ATTRIBUTE_DATA attrs;
-  if (!GetFileAttributesExA(path.c_str(), GetFileExInfoStandard, &attrs)) {
+  if (!GetFileAttributesExW(path.c_str(), GetFileExInfoStandard, &attrs)) {
```
Comment (jonesmz) [inline src/disk_interface.cc:95]: See https://github.com/ninja-build/ninja/pull/1671

File: src/disk_interface.cc [hunk 4]
```diff
@@ -73,13 +90,13 @@ TimeStamp TimeStampFromFileTime(const FILETIME& filetime) {
...
     DWORD win_err = GetLastError();
     if (win_err == ERROR_FILE_NOT_FOUND || win_err == ERROR_PATH_NOT_FOUND)
       return 0;
-    *err = "GetFileAttributesEx(" + path + "): " + GetLastErrorString();
+    *err = "GetFileAttributesExW(" + NarrowPath(path) + "): " + GetLastErrorString();
```
Comment (jonesmz) [inline src/disk_interface.cc:99]: See https://github.com/splinter-build/splinter/blob/10-string_concat/src/string_concat.h

File: src/includes_normalize_test.cc [hunk 5]
```diff
@@ -17,6 +17,7 @@
 #include <algorithm>
 
 #include <direct.h>
+#include <windows.h>
```
Comment (jonesmz) [inline src/includes_normalize_test.cc:20]: #define WIN32_LEAN_AND_MEAN
#define NOMINMAX

File: src/util.cc [hunk 6]
```diff
@@ -147,7 +148,7 @@ bool CanonicalizePath(char* path, size_t* len, uint64_t* slash_bits,
     return false;
   }
 
-  const int kMaxPathComponents = 60;
+  const int kMaxPathComponents = 120;
```
Comment (jonesmz) [inline src/util.cc:151]: Magic number.

File: src/util.cc [hunk 7]
```diff
@@ -345,8 +346,10 @@ int ReadFile(const string& path, string* contents, string* err) {
   // This makes a ninja run on a set of 1500 manifest files about 4% faster
   // than using the generic fopen code below.
   err->clear();
-  HANDLE f = ::CreateFileA(path.c_str(), GENERIC_READ, FILE_SHARE_READ, NULL,
-                           OPEN_EXISTING, FILE_FLAG_SEQUENTIAL_SCAN, NULL);
+
+  HANDLE f = ::CreateFileW(WidenPath(path).c_str(), GENERIC_READ, FILE_SHARE_READ, NULL,
```
Comment (jonesmz) [inline src/util.cc:350]: Calling WidenPath() over and over again on the same string object will result in terrible performance.

See again: https://github.com/ninja-build/ninja/pull/1671

**Comment by javier-martinez-palmer** [comment]
It would be great this is merged and working fine soon. Looking forward to it

**Comment by rpiper** [comment]
Looking forward to this fix as well. I haven't found a way to successfully workaround the long path problem when building with Jenkins.

**Comment by jaeger25** [comment]
Hey all, just wanted to give an update that this is still an issue effecting our team, and I plan to come back and work on getting this to a mergeable state near the end of August / start of September.

**Comment by trzecieu** [comment]
Hello! Is there any obstacle for merging this pull request? Looks quite useful, especially in the context of using ninja by Android Studio on Windows.

**Comment by jhasse** [comment]
closing in favor of #2056

**Comment by Selrak** [comment]
How to use this commit (especially, activate the long path capability) ? (For me with CMake 3.20). Is there an option or something ?

**Comment by SidneyLann** [comment]
still happen in 2025, why I need to find a solution? why not constraint the file length?

**Comment by jhasse** [comment]
> why I need to find a solution?

Because Windows sucks.

**Comment by SidneyLann** [comment]
> > why I need to find a solution?
> 
> Because Windows sucks.

Not related to windows, just use uuid as the fileName should solve the problem!

**Comment by jhasse** [comment]
Right, my bad.

**Comment by h1nson** [comment]
Download the latest Ninja release from the [official repository](https://github.com/ninja-build/ninja/releases)
Replace the existing ninja.exe in $SDK_PATH$\cmake\$CMAKE_VERSION$\bin with the newly downloaded version

**Comment by SidneyLann** [comment]
ok
